Update README to include more info for the CLI (#5)
slack-samples/deno-starter-template: README.md

# Deno Starter Template for Run On Slack apps

This repo is a starter template for new Run On Slack apps. This template uses Deno and the Slack CLI.

Before getting started, make sure you have a development workspace where you have permissions to install apps. If you don’t have one set up, go ahead and [create one](https://slack.com/create).

## Installation

#### Prerequisites

To use this template, you will need to have installed and configured the Slack CLI. 
Do this by following our [Quickstart Guide](https://api.slack.com/future/quickstart).

#### Get Started

Make a Run On Slack app with this with repo by **creating an app from this template**, **configuring your app**, then **writing functions**. 

Once you're done, you can run your app with the CLI's local development server or Deploy your app to production.

### Setup Your Project

```zsh
# Clone this project onto your machine
slack create my-app -t slack-samples/deno-starter-template

# Change into this project directory
cd my-app

> # Run app locally

Author: @misscoded - would love to get your thoughts on including the run + deploy instructions under this heading as well when there's also more in-depth information about how to run and deploy later on in this doc! having at least the run instructions here would match similar templates like bolt JS but is it redundant when we have a dedicated section related to running the app?

Reviewer: I don't think it's redundant, but I could be wrong. Eager developers don't read documentation very well (I am almost always guilty of this), so having a section that very clearly indicates how it works might be useful.

# Deno Starter Template for Run On Slack apps

This repo is a starter template for new Run On Slack apps. This template uses Deno and the Slack CLI.

Before getting started, make sure you have a development workspace where you have permissions to install apps. If you don’t have one set up, go ahead and [create one](https://slack.com/create).

## Template Versions

#### Accessing Different CLI Versions of Templates
This template is compatible with version(s) of the Slack CLI.

To view a template that is compatible with a specific version of the CLI, view the list below:

> - **(current)**[deno-starter-template for CLI v1.0.0](https://github.com/slack-samples/deno-starter-template/tree/main)

Reviewer: Since we've yet to decide on the template versioning approach, let's remove this.

## Installation

#### Prerequisites

To use this template, you will need to have installed and configured the Slack CLI. 
Do this by following our [Quickstart Guide](https://api.slack.com/future/quickstart).

> You will also need a `slack.json` file - you can learn more about this file in the `Project Structure` section.

Reviewer: I don't think this line is necessary since this file is already part of the code they're pulling down

#### Get Started

Make a Run On Slack app with this with repo by **creating an app from this template**, **configuring your app**, then **writing functions**. 

> Once you're done, you can Run your app with the CLI's local development server or Deploy your app to production.

Reviewer: ```suggestion
Once you're done, you can run your app with the CLI's local development server or deploy your app to production.
```

### Setup Your Project

```zsh
# Clone this project onto your machine
slack create my-app -t slack-samples/deno-starter-template

# Change into this project directory
cd my-app

# Run app locally
slack run

# Deploy app
slack deploy
```

> #### Running your app locally

Reviewer: ```suggestion
#### Running Your App Locally
```

While building your app, you can see your changes propagated to your workspace in real-time with `slack run`.

Executing `slack run` starts a local development server, syncing changes to your workspace's development version of your app. (You'll know it's the development version because the name has the string `(dev)` appended).

Your local development server is ready to go when you see the following:

```zsh
Connected, awaiting events

```

When you want to turn off the local development server, use `Ctrl+c` in the command prompt.

> #### Deploying your app to Slack

Reviewer: ```suggestion
#### Deploying Your App to Slack
```

When you're done developing your app, you can deploy it directly to Slack with `slack deploy`.

> #### Configuring your app

Reviewer: ```suggestion
#### Configuring Your App
```

> The first thing we'll do is configure our app's manifest. This will let us configure things like our app's name and the [scopes](https://api.slack.com/scopes) it requires.

Reviewer: If configuring is the first step, should it not come before instructions on how to run the app? Also, shouldn't deploy go very last?

});
```

> #### Writing functions

Reviewer: ```suggestion
#### Writing Functions
```

On our [next-generation platform](https://api.slack.com/future), you can build **Run On Slack functions**, reusable building blocks of automation that are deployed to Slack and accept inputs, perform some calculations, cand provide outputs. 

> Functions can be triggered via Global Shortcut, and we'll be adding support for more function and trigger types in the coming months.

Reviewer: I don't believe this is true anymore, right? They must be link triggers? Since this is likely to change, perhaps we remove this all together.

To create a Run On Slack function:

* **define** the function in the Manifest, then 
* **implement** the function in its respective source file.

> ##### Define your function

Reviewer: ```suggestion
##### Define Your Function
```

</details>

> ##### Implement your function

Reviewer: ```suggestion
##### Implement Your Function
```

A function's implementation must be the default export of the source file. 
When instantiating the function, use `SlackFunctionHandler` and pass 
in your function's `.definition` so that your required inputs and outputs will be 
enforced.

> ## Using the CLI

Reviewer 2: ```suggestion
## Using the Slack CLI
```

To learn more about development with the CLI, you can visit the following guides:

> - [In-depth guide on Creating a New App](https://api.slack.com/future/create)

Reviewer 2: ```suggestion
- [Creating a new app with the CLI](https://api.slack.com/future/create)
```

> - [Configuring an app](https://api.slack.com/future/manifest)

Reviewer 2: ```suggestion
- [Configuring your app](https://api.slack.com/future/manifest)
```

- [Developing locally](https://api.slack.com/future/run)
To view all documentation and guides available for the CLI, visit the [Overview page](https://api.slack.com/future/overview).

When using the CLI to develop your app, here are some helpful commands:

> - `slack version`: Checks current version of your CLI and the SDKs it uses.

Reviewer 2: ```suggestion
- `slack version`: Checks your current version of the Slack CLI.
```
Technically `version` just checks the CLI version

- `slack upgrade`: Checks for updates for current SDK and CLI versions and upgrades them as needed.

> - `slack auth list`: Lists all authorized workspaces.

Reviewer 2: ```suggestion
- `slack auth list`: Lists all workspaces the CLI is logged into.
```
This suggestion for wording is a little awkward so feel free to disregard, but we should probably distinguish that the authorization is for the CLI to work with the workspace (not say, the authorization that might come with an app installation)

- `slack auth login`: Allows you to log in to a new or inactive workspace with the CLI.
To view all other commands available in the CLI, run `slack help`.

## Project Structure

### `manifest.ts`

`manifest.ts` is a configuration for Slack CLI apps using Deno. This file will establish all basic configurations for your application, including app name and description.

### `slack.json`

`slack.json` is a required file for running Slack CLI apps. This file is a way for the CLI to interact with your project's SDK. It defines script hooks which are *executed by the CLI* and *implemented by the SDK.*

An example of this file's format can be found below:
```js
{
  "hooks": {
    "get-hooks": "deno run -q --unstable --allow-read --allow-net https://deno.land/x/deno_slack_hooks@0.0.4/mod.ts",
    // This is a user-defined custom hook that overrides the default "get-manifest"
    "get-manifest": "deno run -q --unstable --config=deno.jsonc --allow-read --allow-net https://deno.land/x/deno_slack_builder@0.0.8/mod.ts --manifest",
    // This is a user-defined custom hook that adds new functionality
    "custom-hook": "deno run my-custom-hook.ts"

>   }

Reviewer: They have the project that contains an example hook: what value does this snippet serve for the developer that's reading? I'm a little concerned that someone will copy/paste these and be getting outdated versions.

}
```

## App Distribution / OAuth

Only implement OAuth if you plan to distribute your application across multiple workspaces.

When using OAuth, Slack requires a public URL where it can send requests. You can use [`ngrok`](https://ngrok.com/download). Check out [this guide](https://ngrok.com/docs#getting-started-expose) for setting it up.

Start `ngrok` to access the app on an external network and create a redirect URL for OAuth. 

```
ngrok http 3000
```

This output should include a forwarding address for `http` and `https` (we'll use `https`). It should look something like the following:

```
Forwarding   https://3cb89939.ngrok.io -> http://localhost:3000
```

Navigate to **OAuth & Permissions** in your app configuration and click **Add a Redirect URL**. The redirect URL should be set to your `ngrok` forwarding address with the `slack/oauth_redirect` path appended. For example:

```
https://3cb89939.ngrok.io/slack/oauth_redirect

> ```

Reviewer: I don't think this section is relevant at all?
